### PR TITLE
Autocomplete: Reopen @-mention typeahead when editing inside `**`

### LINKED ISSUE #1967: After choosing an @-mention option, backspacing should reopen autocomplete
In Web, if an @-mention autocomplete option is chosen, pressing backspace will reopen the autocomplete. For example, if the chosen option puts `@**Chris Bobbe**` in the compose box and then backspace is pressed until it becomes something like `@**Chris`, the autocomplete will be shown for all the matching results of `Chris`, which means ignoring the starting `**` from the query. We implemented a similar behavior for the channel link autocomplete in #1902, but this issue is for doing so in @-mention autocomplete.

### PR DESCRIPTION
**Pull request by MritunjayTiwari14**
Fixes #1967.

<details>
  <summary>Changes</summary>

| Before | After |
| :---: | :---: |
| <video src="https://github.com/user-attachments/assets/b9a97d53-d4d5-4122-86af-79be595089f9" width="350" controls></video> | <video src="https://github.com/user-attachments/assets/042f2d7c-03d7-4a74-bf2b-247d2cfbd5c6" width="350" controls></video> |

</details>

### CODE REVIEW

**Comment by MritunjayTiwari14** [comment]
@chrisbobbe Ready for review.

File: lib/model/autocomplete.dart
Comment (Khader-1) [inline lib/model/autocomplete.dart]: I think this should be bumped to be `+ 3`. The longest prefix after `@` is now `_**` (3 chars, for a silent mention like @_**foo). With a small `maxChannelNameLength`, the lookback limit can prevent the scanner from reaching the `@` and then the intent would be missed.
 
For example:
 `@_**ab` with `maxChannelName`: 1, lookback is 1+2+2=5 but the pattern is 6 chars.

File: lib/model/autocomplete.dart [hunk 1]
```diff
@@ -115,15 +115,24 @@ final RegExp _mentionIntentRegex = (() {
...
     beforeAtSign
     + r'@(_?)' // capture, so we can distinguish silent mentions
-    + r'(|'
-      // Reject on whitespace right after "@" or "@_". Emails can't start with
-      // it, and full_name can't either (it's run through Python's `.strip()`).
-      + r'[^\s' + fullNameAndEmailCharExclusions + r']'
-      + r'[^'   + fullNameAndEmailCharExclusions + r']*'
+    + r'(?:'
+      // Reject on whitespace and "**" right after "@" or "@_". Emails can't start
+      // with both, and full_name can't either (it's run through Python's `.strip()`).
```
Comment (Khader-1) [inline lib/model/autocomplete.dart]: (nit):
```suggestion
      // with either, and full_name can't either (it's run through Python's `.strip()`).
```

File: test/model/autocomplete_test.dart [hunk 2]
```diff
@@ -134,10 +134,51 @@ void main() {
...
+    doTest('Ask @**Chris`Bobbe^', null);
+    doTest('Ask @**Chris\\^', null);
+    doTest('Ask @**Chris\\Bobbe^', null);
+    doTest('Ask @**Chris>^', null);
+    doTest('Ask @**Chris>Bobbe^', null);
+    doTest('Ask @**Chris"^', null);
+    doTest('Ask @**Chris"Bobbe^', null);
+
+    // "**" are not allowed inside the query.
+    doTest('Ask @**Chris**Bobbe^', null);
```
Comment (Khader-1) [inline test/model/autocomplete_test.dart:193]: I'd add more edge cases here, maybe:
```dart
// Three asterisks: "@***" = prefix "@**" + query "*".
doTest('Ask ~@***^', mention('*'));
doTest('Ask ~@_***^', silentMention('*'));
doTest('~@_**ab^', silentMention('ab'), maxChannelName: 1);
doTest('~@**abc^', mention('abc'), maxChannelName: 1);
doTest('~@**ab^', mention('ab'), maxChannelName: 1);
```

**Review by Khader-1 — COMMENTED** [review]
Thanks @MritunjayTiwari14! I think this is close, a few notes bellow.

File: lib/model/autocomplete.dart
Comment (MritunjayTiwari14) [reply lib/model/autocomplete.dart]: Yes, it does clearly make sense. This should be increased to 3 because of `_**`. Thanks for noticing!

File: test/model/autocomplete_test.dart [hunk 3]
```diff
@@ -134,10 +134,83 @@ void main() {
...
+
+    // Three asterisks but fully closed.
+    doTest('Ask ~@***^**', mention('*'));
+    doTest('Ask ~@_***^**', silentMention('*'));
+
+    // Four asterisks: "@****" = prefix "@**" + query "**" (which is invalid).
+    doTest('Ask @****^', null);
+    doTest('Ask @_****^', null);
+
+    // " ' " (Aspotrophe) and "-" are allowed inside the query.
```
[the remark this answers is not in the document]
Comment (MritunjayTiwari14) [reply test/model/autocomplete_test.dart]: Thanks!

**Comment by MritunjayTiwari14** [comment]
@chrisbobbe Waiting for your review on this...

File: lib/model/autocomplete.dart
Comment (chrisbobbe) [inline lib/model/autocomplete.dart:30]: The reasoning in https://github.com/zulip/zulip-flutter/pull/2166#discussion_r2900319213 for bumping the 2 to a 3 doesn't quite hold, I think. The max length of a channel name isn't relevant when you're trying to do a silent @-mention, because in that case you're searching for users, not channels.

I think the complexity of trying to maintain an exactly correct number here isn't worth it, and we should just do the simple thing of hard-coding it to 30 or 60 or something. This came up in another context recently: https://github.com/zulip/zulip-flutter/pull/2137#discussion_r2950316709

File: lib/model/autocomplete.dart [hunk 4]
```diff
@@ -73,7 +73,7 @@ extension ComposeContentAutocomplete on ComposeContentController {
       if (charAtPos == '@') {
         final match = _mentionIntentRegex.matchAsPrefix(textUntilCursor, pos);
         if (match == null) continue;
-        query = MentionAutocompleteQuery(match[2]!, silent: match[1]! == '_');
+        query = MentionAutocompleteQuery(match[2] ?? match[3]!, silent: match[1]! == '_');
```
Comment (chrisbobbe) [inline lib/model/autocomplete.dart]: If I understand correctly, I think this could be simplified by using duplicate named capture groups? I see an existing TODO mentioning that feature:

```dart
  // TODO(upstream): maybe use duplicate-named capture groups for better readability?
  //   https://github.com/dart-lang/sdk/issues/61337
```

and that Dart PR was merged recently! 🎉 Can we try that?

File: test/model/autocomplete_test.dart [hunk 5]
```diff
@@ -134,10 +134,83 @@ void main() {
...
+
+    // "**" are not allowed inside the query.
+    doTest('Ask @**Chris**Bobbe^', null);
+
+    // Three asterisks: "@***" = prefix "@**" + query "*".
+    doTest('Ask ~@***^', mention('*'));
+    doTest('Ask ~@_***^', silentMention('*'));
+    doTest('~@_**ab^', silentMention('ab'), maxChannelName: 1);
+    doTest('~@**abc^', mention('abc'), maxChannelName: 1);
+    doTest('~@**ab^', mention('ab'), maxChannelName: 1);
```
Comment (chrisbobbe) [inline test/model/autocomplete_test.dart:197]: Is there a reason to allow "*" in the query at all? We have a comment suggesting that no users will ever match if that character (or a handful of others) is in the query:

```dart
  // Characters that would defeat searches in full_name and emails, since
  // they're prohibited in both forms. These are all the characters prohibited
  // in full_name except "@", which appears in emails. (For the form of
  // full_name, find uses of UserProfile.NAME_INVALID_CHARS in zulip/zulip.)
  const fullNameAndEmailCharExclusions = r'\*`\\>"\p{Other}';
```

**Review by chrisbobbe — COMMENTED** [review]
Thanks, and I'm sorry I've let this sit.

Comments below, including one that unlocks some simplification, I think: if we can continue prohibiting stars ("*") in the query, then I think we can neatly express three mutually exclusive branches in the regex:
- no stars before or after the query
- two stars before the query and no stars after
- two stars before the query and one star after (for the case where you've backspaced through one star after finishing an autocomplete)

Would you give that a try?

File: lib/model/autocomplete.dart (hunk 4 above)
Comment (MritunjayTiwari14) [reply lib/model/autocomplete.dart]: sure!

File: lib/model/autocomplete.dart
Comment (MritunjayTiwari14) [reply lib/model/autocomplete.dart:30]: Well, I disagree, maxlookback should be bumped from 2 to 3. I totally agree, channel name isn't relevant. But how the channel name is affected when the value is bumped from 2 to 3 is what I can't understand in your comment.

Those two are totally independent things, when we introduced silent mention. The min length would inflate from being `**`, i.e., 2, to `_**`, i.e., 3?

**Comment by MritunjayTiwari14** [comment]
Thanks, @chrisbobbe, for the review. I have left a [comment](https://github.com/zulip/zulip-flutter/pull/2166#discussion_r3067697539) on your feedback. As well as addressing all of your feedback. PTAL.

## Changes:
- Used a 3 mutually exclusive case-wise breakdown to identify both normal and silent mentions. Please check the commit description for more details.
- Added a post-refactor commit to replace hardcoded RegExp group capture using `namedGroup()`.
- I have not updated the maxlookback value since the last review; it's still (4 + 2*maxChannelName).
